helix-core/src/main/java/org/apache/helix/controller/stages/PersistAssignmentStage.java

```diff
@@ -128,25 +130,30 @@ private void persistAssignment(final Resource resource, final ResourceController
       if (assignmentToPersist != null && hasInstanceMapChanged(assignmentToPersist, idealState)) {
         for (Partition partition : assignmentToPersist.keySet()) {
           Map<String, String> instanceMap = assignmentToPersist.get(partition);
-          idealState.setInstanceStateMap(partition.getPartitionName(), instanceMap);
+          delta.setInstanceStateMap(partition.getPartitionName(), instanceMap);
         }
         needPersist = true;
       }
 
       if (needPersist) {
         // Update instead of set to ensure any intermediate changes that the controller does not update are kept.
-        accessor.updateProperty(keyBuilder.idealStates(resourceId), new DataUpdater<ZNRecord>() {
-          @Override
-          public ZNRecord update(ZNRecord current) {
-            if (current != null) {
-              // Overwrite MapFields and ListFields items with the same key.
-              // Note that default merge will keep old values in the maps or lists unchanged, which is not desired.
+        accessor.updateProperty(keyBuilder.idealStates(resourceId), current -> {
+          if (current != null) {
+            ZNRecord deltaRecord = delta.getRecord();
+            // Overwrite MapFields and ListFields items with the same key.
+            if (!deltaRecord.getMapFields().isEmpty()) {
+              // Note that default merge will keep old values in the maps unchanged, which is not desired.
               current.getMapFields().clear();
-              current.getMapFields().putAll(idealState.getRecord().getMapFields());
-              current.getListFields().putAll(idealState.getRecord().getListFields());
+              current.getMapFields().putAll(deltaRecord.getMapFields());
+            }
+            if (!deltaRecord.getListFields().isEmpty()) {
```

Comment:
Question: why we need this check? We can add lists field anyway just not clear it up, right?

Reply:
This is for backward compatibility. Not adding it will be fine, logic-wise. But it will change the behavior, so some tests fail. This is because the change in this PR changes the input.

Since adding this check also makes sense to me, I decide to keep the same behavior.
What's your opinion?

Reply:
I agree better to keep the behavior. But the original code does not perform the check, right? The check changed the behavior of our code that only list field contains something then we add. 

Either way is fine. I dont have strong preference because the only thing I can think about is custom mode / non rebalanced resource does not have the list field.  For non rebalanced resource, it will be eventually caught up if the it get rebalanced with list field fulfilled.

Reply:
Since I changed the logic, if I pass an empty map through the "deltaRecord", it means don't update. Originally it means the IdealState should have the empty map. So the meaning of the input has been changed. So this check will ensure the behavior keeps the same.